### DISCUSSION TITLE
Adding Brotli Perl modules for exiftool to find

### DISCUSSION THREAD

**Discussion by hsivonen** [Homebrew/discussions#5013] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.1
ORIGIN: https://github.com/Homebrew/brew
HEAD: 7817518992920b8dbd15e48a5e6c5ad768f4ba8b
Last commit: 6 days ago
Core tap JSON: 30 Dec 14:24 UTC
Core cask tap JSON: 30 Dec 14:24 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: 15.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

When invoking [`exiftool`](https://formulae.brew.sh/formula/exiftool) 12.70 on a JPEG XL image generated with the [`cjxl`](https://formulae.brew.sh/formula/jpeg-xl) tool (which compresses metadata with Brotli by default), `exiftool` says (among other output):
"Install IO::Uncompress::Brotli to decode Brotli-compressed metadata"

Installing the required pod from CPAN to the default location (`/Library/Perl/5.30/darwin-thread-multi-2level/`) does not help. Making a symlink from `/opt/homebrew/Cellar/exiftool/12.70/libexec/lib/IO` to `/Library/Perl/5.30/darwin-thread-multi-2level/IO` does not help.

What's the right way to install the requisite Perl module so that `brew`-installed `exiftool` will find it?

**Comment by cho-m** [comment]
`exiftool` should be using macOS Perl. I think Sonoma has some remnants of Perl 5.30 along with Perl 5.34. We use latter (i.e. shebang will point to `/usr/bin/perl5.34`).

Maybe try reinstalling pod with `/usr/bin/cpan5.34`? It looks like your default location was for Perl 5.30.

**Reply by Qhilm** [reply]
Installing `IO::Uncompress::Brotli` with the macOS `/usr/bin/cpan` fails when running as user, one needs to `sudo cpan` it seems (permission error when attempting to create a directory). I'm not a huge fan of giving cpan root permissions I must admit, I'd much rather find a way to make exiftool use brew's perl, maybe using PATH.

Seeing libvips 8.16 will support writing exif and ICC in JPEG XL files, I'm not sure that it is worth investing much time in this.

**Reply by Qhilm** [reply]
hmmm, addendum, libvips 8.16 will support writing EXIF to jxl, but it will not support loading EXIF from TIFF, only from jpeg files.

Touching the macOS Perl is still a bad idea in my view. There should be a way to tell exiftool to use the brew perl.

I installed perl using brew, I installed "IO::Uncompress::Brotli" using cpan, I added `eval "$(perl -I$HOME/perl5/lib/perl5 -Mlocal::lib=$HOME/perl5)"` to my `.zshrc` and checked using `env | grep PERL` that the following is defined:
```
PERL5LIB=/Users/myuser/perl5/lib/perl5
PERL_LOCAL_LIB_ROOT=/Users/myuser/perl5
PERL_MB_OPT=--install_base "/Users/myuser/perl5"
PERL_MM_OPT=INSTALL_BASE=/Users/myuser/perl5
```

but still, exiftool is complaining about the missing `IO::Uncompress::Brotli` module.

Not sure what else to try.

[edit] I checked that `/Users/myuser/perl5/lib/perl5/darwin-thread-multi-2level/IO/Uncompress/Brotli.pm` exists.

**Reply by cmahnke** [reply]
Maybe there should be an issue on this? Since JXL is a first class citizen in recent Mac OS this should be included as a hard dependency...

I just realized I had a similar issue before:
https://github.com/Homebrew/homebrew-core/issues/192273

A naive approach won't work either:
```
brew install perl
/opt/homebrew/bin/cpan IO::Uncompress::Brotli
```

This solved ist for me:
```
sudo /usr/bin/cpan IO::Uncompress::Brotli
ln -s /Library/Perl/5.34/darwin-thread-multi-2level/IO  /opt/homebrew/Cellar/exiftool/12.76/libexec/lib
```

But this solution is quite suboptimal:
* `exiftool` shouldn't depend on the system `perl` installation in order to avoid the need tinkering with it. The current (rather poor) choice is likely due to trying to reduce dependencies
* This certainly won't survive an update of `exiftool`
* * There needs to be a automatic way for `exiftool` to use a homebrew `perl` installation, this can be done by updating the file in the formula.

**Reply by Qhilm** [reply]
Thanks a lot for this. I agree it's an issue and I fully agree that messing around with the OS perl cannot be considered a solution.

Would be interesting to see how other software depending on a specific perl module deal with this dependency. I can't find anything so far, but my search fu in this area is quite limited.

**Reply by cmahnke** [reply]
I've added [#195911](https://github.com/Homebrew/homebrew-core/issues/195911)

**Reply by Qhilm** [reply]
> This solved ist for me:
> 
> ```
> sudo /usr/bin/cpan IO::Uncompress::Brotli
> ln -s /Library/Perl/5.34/darwin-thread-multi-2level/IO  /opt/homebrew/Cellar/exiftool/12.76/libexec/lib
> ```

Does not work for me, for what it's worth (macOS 15.0.1).

I get "loadable library and perl binaries are mismatched (got handshake key 0x10200080, needed 0xfb80080)" when I run `exiftool -tagsFromFile`

**Reply by cmahnke** [reply]
That look like you're not using the bundled / system perl to run `exiftool`, maybe there are environment variables that change the used interpreter?
You can also try to link the module from the library path that your interpreter uses, it might be somewhere below `/opt/homebrew/Cellar/perl/5.40.0/lib/perl5/` (given it's installed there)

**Reply by Qhilm** [reply]
Indeed, I had even written it above, I forgot about it:

```
PERL5LIB=/Users/myuser/perl5/lib/perl5
PERL_LOCAL_LIB_ROOT=/Users/myuser/perl5
PERL_MB_OPT=--install_base "/Users/myuser/perl5"
PERL_MM_OPT=INSTALL_BASE=/Users/myuser/perl5
```

I removed these and it worked, just had to add `-ignoreMinorErrors`. I don't understand the "Wrapped JXL codestream in ISO BMFF container" error, but it's ingnored with this flag.

**Reply by cmahnke** [reply]
It's just a warning, as you wrote...
I've seen it before, maybe this goes away if you run a dev version of `exiftool`. Note, that the stable (production) release is always a bit behind, see https://exiftool.org/history.html

**Reply by AngelaDMerkel** [reply]
> This solved ist for me:
> 
> ```
> sudo /usr/bin/cpan IO::Uncompress::Brotli
> ln -s /Library/Perl/5.34/darwin-thread-multi-2level/IO  /opt/homebrew/Cellar/exiftool/12.76/libexec/lib
> ```

This also solved it for me, but I am also not a fan of having to install using sudo
